### PR TITLE
[PF-576] Add Harness action catalog surface

### PR DESCRIPTION
**Pull request by mbenhamd**
## Summary

- Adds `session.actions.list/search/refresh()` for the local Harness v1 desktop action catalog.
- Aggregates code/workspace skill action metadata and MCP tool descriptors without adding execution or lifecycle controls.
- Keeps workspace skill discovery metadata-only, caches action catalog reads, supports explicit refresh, bounds MCP cache keys by context, negative-caches unhealthy MCP servers briefly, and aborts MCP catalog requests on timeout.
- Adds Harness reference docs and a minor changeset for `@mastra/core`.

Linear: https://linear.app/papersflow/issue/PF-576/harness-v1desktop-unified-action-catalog-search-surface

## Validation

- `pnpm --filter ./packages/core test:unit src/harness/v1/session.actions.test.ts src/harness/v1/session.skills.test.ts src/harness/v1/session.mcp.test.ts`
- `pnpm build:core`
- `git diff --check`
- Local Codex review passes: correctness/API/security/scope and tests/maintainability/merge readiness
- CodeRabbit CLI: final rerun completed with no findings

## Notes

- No PostgreSQL/Docker storage tests were run because this PR does not touch storage adapters, persistence, or PG paths.

## ELI5
This PR adds a read-only, searchable "menu" of actions to Harness sessions so clients can list and search actions from workspace skills and connected MCP tools without running them. It caches results for speed and lets callers refresh the catalog when skills or tools change.

## Overview
Adds a new Harness v1 in-memory session surface: session.actions with three methods:
- list(options?): Return catalog entries (skill metadata + MCP tool descriptors) with filtering, pagination, and clone-safe snapshots.
- search(query, options?): Convenience search across the catalog.
- refresh(): Clear caches and force catalog rediscovery.

Catalog entries are descriptive only (no execution or permission granting); hosts should route execution via existing skill/tool execution APIs (e.g., session.skills.use()). MCP discovery used for the catalog is workspace-aware but will not provision a workspace solely to build the catalog.

## Key Implementation Details
- New public types and re-exports for catalog entries and options.
- Per-session caching and single-flight discovery for skill-derived entries; per-MCP-server caching and single-flight for MCP tool listings.
- MCP robustness:
  - Abortable MCP requests via timeout/AbortSignal to avoid hangs.
  - Short-term negative caching for failed/timed-out MCP servers to reduce repeated failures.
  - MCP cache keys scoped by workspace context; catalog-only reads can skip eager workspace provisioning.
  - Request-context propagation to MCP calls.
- session.skills.refresh() clears action-catalog caches and resolves/clears in-flight promises.
- list/search validate and normalize inputs, filter by source/query, apply offset/limit, sort by source priority then stable id, and return cloned projection-safe entries.
- Methods reject with HarnessSessionClosedError after session.close().
- Tests: timing-based assertions were stabilized (commit 774fb3cfcb) and an extensive Vitest suite covers catalog merging, caching, workspace-scoped MCP behavior, timeout/negative-caching, cloning safety, validation, and session-closed behavior.

## Files Changed (high level)
- Documentation
  - docs/src/content/en/reference/harness/harness-class.mdx (+16): New "Session action catalog" docs and examples.
  - .changeset/pf-576-action-catalog.md (+13): Changeset entry for `@mastra/core` (minor).
- Public API surface
  - packages/core/src/harness/v1/types.ts (+77): Catalog types (entries, sources, list options).
  - packages/core/src/harness/v1/index.ts (+6): Re-exports new types.
- Implementation
  - packages/core/src/harness/v1/session.ts (+~509/-10): Adds readonly actions (list/search/refresh), caching, MCP timeout/negative-cache behavior, workspace-aware MCP keys, and related internal API changes (_mcpListTools abortSignal/resolveWorkspace and _buildRequestContext resolveWorkspace).
- Tests
  - packages/core/src/harness/v1/session.actions.test.ts (+~864): Expanded Vitest suite covering catalog behavior and edge cases.
  - packages/core/src/workspace/skills/workspace-skills.test.ts (+39): Added refresh and dynamic-path resolution tests.
  - packages/core/src/workspace/skills/workspace-skills.ts (+8/-1): Path-resolution caching tweaks.

## Validation / Notes
- Unit tests for the affected harness tests passed locally per author (multiple runs reported).
- Build: pnpm build:core.
- git diff --check passed.
- Local reviews and CodeRabbit CLI reported no findings.
- PR does not modify storage adapters or PostgreSQL paths; an isolated Postgres check was run by the author but PG/Docker storage tests are not required for this change.
- Linear ticket: PF-576.

[Review Change Stack](https://app.coderabbit.ai/change-stack/mbenhamd/mastra/pull/162?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/core/src/harness/v1/session.ts [hunk 1]
```diff
@@ -1448,6 +1601,285 @@ export class Session {
...
+    };
+  }
+
+  private _actionMcpCatalogCacheKey(server: HarnessMcpServerDescriptor): string {
+    return [
+      server.key,
+      this._record.modeId,
+      this._record.modelId ?? '',
+      this._workspace?.id ?? '',
+    ].join('\0');
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/harness/v1/session.ts]: Agent-authored reply with user authorization. I verified this against the current PR head e9957b56cd.

The concern was valid for the previous version, but the pushed fix changes the action-catalog path so MCP action discovery calls `_mcpListTools(server.key, abortSignal, { resolveWorkspace: false })`. In `_buildRequestContext`, `resolveWorkspace: false` uses the already-materialized `this._workspace` only and does not call `_getWorkspaceUnchecked()`, so the cache key built from `this._workspace?.id ?? ""` matches the workspace context used for the fetch.

Evidence added in `packages/core/src/harness/v1/session.actions.test.ts`:
- `uses the resolved workspace identity for MCP action cache keys when available`
- `does not provision or block behind a slow workspace provider for MCP action discovery`
- `does not repopulate MCP action caches from discovery cleared by refresh`

Local validation after the fix:
- `pnpm --filter ./packages/core test:unit src/harness/v1/session.actions.test.ts src/harness/v1/session.skills.test.ts src/harness/v1/session.mcp.test.ts`
- `pnpm build:core`
- `git diff --check`
- two local Codex review passes, both no findings

I believe this thread is resolved by e9957b56cd.

File: packages/core/src/harness/v1/session.ts [hunk 2]
```diff
@@ -1448,6 +1601,285 @@ export class Session {
...
+      .slice(normalized.offset, normalized.offset + normalized.limit)
+      .map(entry => cloneActionCatalogEntry(entry));
+  }
+
+  private async _actionsRefresh(): Promise<void> {
+    this._assertLive('actions.refresh()');
+    this._actionsSkillEntriesCache = undefined;
+    this._actionsSkillEntriesResolving = undefined;
+    this._actionsMcpEntriesCacheByServer.clear();
+    this._actionsMcpEntriesResolvingByServer.clear();
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/harness/v1/session.ts]: Agent-authored reply with user authorization.

Addressed in `9f1e669c4a` (`[PF-576] Refresh workspace skills for action catalog`). `session.actions.refresh()` now refreshes the already-materialized workspace skill source before rebuilding the action catalog and clears both the skill catalog and action catalog caches before/after refresh so failed refreshes cannot leave stale entries pinned.

I also fixed the underlying first-refresh path handling in `WorkspaceSkillsImpl.refresh()`: refresh now resolves skill paths before first discovery, using an explicit `#pathsResolved` sentinel so a context-resolved empty path list remains a valid cached result and is not accidentally re-resolved without that context.

Added coverage:
- `refreshes an already-materialized workspace skill source before rebuilding skill actions`
- `should discover skills when refresh is called before first list`
- `should preserve context-resolved empty paths across refresh`

Validation after the patch: focused core Vitest suite (`session.actions`, `session.skills`, `session.mcp`, `workspace-skills`) passed with 139 tests; `git diff --check` passed; `pnpm build:core` passed; two local Codex review passes found no issues; CodeRabbit CLI retry completed with no findings. This thread should be resolved after this patch.

**Comment by mbenhamd** [comment]
Agent-authored update with user authorization.

Addressed CodeRabbit review `4339745240` / nitpick about wall-clock timing in `session.actions.test.ts` in `774fb3cfcb` (`[PF-576] Stabilize action catalog provider test`). The slow workspace provider test no longer asserts `<500ms`; it now uses deterministic `createCalls` / `resumeCalls` counters and asserts the provider is never invoked while MCP action discovery still succeeds and caches the MCP tool list.

Validation for this final test cleanup:
- `pnpm --filter ./packages/core test:unit src/harness/v1/session.actions.test.ts src/workspace/skills/workspace-skills.test.ts` passed: 2 files, 114 tests, no type errors.
- second local Codex review independently ran `pnpm --filter ./packages/core test -- src/harness/v1/session.actions.test.ts`; passed: 12 tests, no type errors.
- `git diff --check` passed.
- two local Codex review passes reported no findings.
- CodeRabbit CLI completed with no findings on the previous full diff; the final test-only retry ended during Reviewing without producing a report, so I am not counting that retry as pass evidence.

PG/Docker note: PF-576 does not touch PG/storage adapter code, so I did not start an isolated Postgres Docker test for this PR. For the next PG/storage Harness issue I will run it via an isolated Docker container/port/volume so it does not interfere with other agents.

**Comment by mbenhamd** [comment]
dev1 evidence update: after the PG/Docker reminder I ran an extra isolated Harness PG domain check even though PF-576 does not touch PG/storage adapter code.

Isolation details:
- One-off container: `pf-harness-pg-dev1-20260521-162`
- Image: `pgvector/pgvector:0.8.2-pg16`
- Host bind: `127.0.0.1:55432 -> 5432`, avoiding the repo default `5434` / compose `pg-test-db`
- Passed command: `POSTGRES_HOST=127.0.0.1 POSTGRES_PORT=55432 POSTGRES_DB=mastra POSTGRES_USER=postgres POSTGRES_PASSWORD=postgres pnpm --dir /tmp/mbenhamd-mastra --filter @mastra/pg exec vitest run src/storage/domains/harness/index.test.ts`
- Result: 1 file passed, 12 tests passed
- Cleanup: `docker stop pf-harness-pg-dev1-20260521-162`; rechecked `docker ps`, the container is gone

Note: a first attempt with only `DB_URL=...:55432` failed because `stores/pg/src/storage/test-utils.ts` reads `POSTGRES_*` env vars and defaulted to `5434`. The passing command above is the correct non-interfering PG path for this package.

**Comment by mbenhamd** [comment]
Agent-authored evidence update with user authorization.

Final head: `ee6d08e4472f4493946b2c6cae4a371564eea9bf`.

Resolved review feedback in this final push:
- MCP timeout fallback no longer creates permanent empty cache entries; hung discovery uses finite backoff and bounded retry.
- `skills.refresh()` and `actions.refresh()` refresh already-materialized workspace skill sources and clear skill/action caches before/after refresh.
- Same-name workspace skill actions now expose canonical `source.ref` so hosts can execute the unambiguous `session.skills.use(action.source.ref)` path; path-addressable workspace actions are no longer hidden by same-name code actions.
- Successful MCP action catalog cache entries now have a short TTL so context-filtered MCP catalogs can refresh instead of staying pinned indefinitely.
- Clone failures on action catalog entries now fail explicitly instead of casting `undefined`.

Local validation run on current head:
- `pnpm --filter ./packages/core run test:unit src/harness/v1/session.actions.test.ts src/harness/v1/session.skills.test.ts src/harness/v1/session.mcp.test.ts src/workspace/skills/workspace-skills.test.ts` — 4 files, 146 tests passed.
- `pnpm --filter ./packages/core exec prettier --check src/harness/v1/session.ts src/harness/v1/session.actions.test.ts src/harness/v1/types.ts ../../docs/src/content/en/reference/harness/harness-class.mdx` — passed.
- `pnpm --filter ./packages/core exec eslint src/harness/v1/session.ts src/harness/v1/session.actions.test.ts src/harness/v1/types.ts src/workspace/skills/workspace-skills.ts` — passed.
- `pnpm --filter ./packages/core check` — passed.
- `pnpm build:core` — passed.
- `git diff --check origin/main` — passed.
- Docker Postgres harness storage test: isolated container `pf-harness-pg-dev1-20260522-576l`, bound only to `127.0.0.1:55486`, tmpfs data dir, ran `POSTGRES_HOST=127.0.0.1 POSTGRES_PORT=55486 POSTGRES_DB=mastra POSTGRES_USER=postgres POSTGRES_PASSWORD=postgres timeout 360 pnpm --filter @mastra/pg exec vitest run src/storage/domains/harness/index.test.ts` — 1 file, 13 tests passed. Container removed after the run; `docker ps`/`ss -ltnp` confirmed no `55486` listener remained.
- Local Codex review pass 1 after final fixes — no findings.
- Local Codex review pass 2 after final fixes could not run because Codex CLI hit the usage limit until 19:46; not counted as completed.
- CodeRabbit CLI `coderabbit review --agent --base origin/main` — completed with 0 findings.

Post-push GitHub evidence after the 3-minute wait:
- `gh pr checks 162 --repo mbenhamd/mastra` — all non-skipped checks/statuses green, including CodeRabbit, Docs E2E tests, Lint docs, redirects, and contributor action statuses.
- Review threads query — all threads resolved and outdated.
- PR is mergeable at the head above.

No production Convex, production Docker, secrets, signing, or release/publish actions touched.
